torch/csrc/utils/out_types.cpp

```diff
@@ -0,0 +1,37 @@
+#include <torch/csrc/utils/out_types.h>
+
+namespace torch {
+namespace utils {
+
+void check_out_type_matches(const at::Tensor& result,
+                            at::ScalarType scalarType, bool scalarType_is_none,
+                            c10::optional<at::Layout> layout,
+                            const at::Device& device, bool device_is_none) {
+  if (scalarType_is_none && !layout && device_is_none) {  // common case
+    return;
+  }
+  if (!scalarType_is_none && result.scalar_type() != scalarType) {
+    AT_ERROR(
```

Comment:
AT_ERROR is deprecated in favor of TORCH_CHECK (for errors that a user input can trigger) or TORCH_INTERNAL_ASSERT (for internal logic failures that should never occur).